### ISSUE TITLE
hw3: about func tf.stop_gradient()

### ISSUE THREAD

**Issue by JerryHan-Peking** [PeiqinSun/tf-tutorials#56]
code: attack.py, line 22, **x_round = x_clip + tf.stop_gradient(x_clip // 1 - x_clip)**
I have 2 questions.

A) Can I change it as
**x_round = x_clip//1
tf.stop_gradient(x_clip)** ?

B) What is the function **tf.stop_gradient(input)** doing?
In function **tf.gradients(ys,xs)** , there are arguments xs, ys for computing _d ys /d xs_.
When input=x_clip // 1 - x_clip, is "x_clip // 1 - x_clip" a "ys" ? And then where is the "xs"?

**Comment by chu-tianxiang** [comment]
Floor division is not continuous and thus not differentiable,  so it is approximated using a continuous surrogate function _(x_clip)_ plus a small discontinuous function _(x_clip // 1 -x_clip)_.  When back prop, we take the gradient of surrogate function as replacement for original one, _stop_gradient_ simply means this part is ignored in gradient computation and backprop.
